processing/src/main/java/org/apache/druid/segment/IndexIO.java

```diff
@@ -682,6 +689,102 @@ public Metadata getMetadata()
       return index;
     }
 
+    private Map<String, Supplier<ColumnHolder>> readProjectionColumns(
+        ObjectMapper mapper,
+        SegmentLazyLoadFailCallback loadFailed,
+        AggregateProjectionSpec projectionSpec,
+        SmooshedFileMapper smooshedFiles,
+        Map<String, Supplier<ColumnHolder>> columns,
+        Interval dataInterval
+    ) throws IOException
+    {
+      final Map<String, Supplier<ColumnHolder>> projectionColumns = new LinkedHashMap<>();
+      for (DimensionSchema groupingColumn : projectionSpec.getGroupingColumns()) {
+        final String smooshName = Projections.getProjectionSmooshV9FileName(projectionSpec, groupingColumn.getName());
+        final ByteBuffer colBuffer = smooshedFiles.mapFile(smooshName);
+
+        if (columns.containsKey(groupingColumn.getName())) {
+          final String internedColumnName = SmooshedFileMapper.STRING_INTERNER.intern(groupingColumn.getName());
+          projectionColumns.put(internedColumnName, Suppliers.memoize(
+              () -> {
+                try {
+                  ColumnDescriptor serde = mapper.readValue(SERIALIZER_UTILS.readString(colBuffer), ColumnDescriptor.class);
+                  return serde.readProjection(colBuffer, columnConfig, smooshedFiles, columns.get(internedColumnName).get());
+                }
+                catch (IOException | RuntimeException e) {
+                  log.warn(e, "Throw exceptions when deserialize column [%s].", groupingColumn.getName());
+                  loadFailed.execute();
+                  throw Throwables.propagate(e);
+                }
+              }
+          ));
+        } else {
+          registerColumnHolder(
+              true,
+              projectionColumns,
+              groupingColumn.getName(),
+              mapper,
+              colBuffer,
+              smooshedFiles,
+              loadFailed
+          );
+        }
+
+        if (groupingColumn.getName().equals(projectionSpec.getTimeColumnName())) {
+          projectionColumns.put(ColumnHolder.TIME_COLUMN_NAME, projectionColumns.get(groupingColumn.getName()));
+        }
+      }
+      for (AggregatorFactory aggregator : projectionSpec.getAggregators()) {
+        final String smooshName = Projections.getProjectionSmooshV9FileName(projectionSpec, aggregator.getName());
+        final ByteBuffer aggBuffer = smooshedFiles.mapFile(smooshName);
+        registerColumnHolder(
+            true,
+            projectionColumns,
+            aggregator.getName(),
+            mapper,
+            aggBuffer,
+            smooshedFiles,
+            loadFailed
+        );
+      }
+      if (projectionSpec.getTimeColumnName() == null) {
+        // todo (clint): get the count column name from project spec since if it was doing things correctly it might
+        //  possibly not be __count
+        // todo (clint): also, why do we still have to read columns just to know the stupid row count
```

Comment:
obvious answer- put the row count in Metadata ;)

Reply:
looking into this, but it does mean we need some sort of wrapper in the `Metadata` projections so that we can still intern the schema part of it

Reply:
did this, i split `AggregateProjectionMetadata` out of `AggregateProjectionSpec` (the api type). This metadata type has the schema which is interned, and the row count. The comparator for `AggregateProjectionMetadata` considers the row count first, so the smallest row counts go first (row count aren't available for incremental index so it just has zeros and falls back to pure schema comparison)